### PR TITLE
add support pending operation marker

### PR DESCRIPTION
**Pull request by suwatch**

### CODE REVIEW

File: Kudu.Core/Helpers/PostDeploymentHelper.cs [hunk 1]
```diff
@@ -394,6 +394,51 @@ public static void RunPostDeploymentScripts(TraceListener tracer)
...
+
+            if (timeout <= TimeSpan.Zero || timeout >= TimeSpan.FromMinutes(DefaultTimeoutMinutes))
+            {
+                // track at most N mins by default
+                timeout = TimeSpan.FromMinutes(DefaultTimeoutMinutes);
+            }
+
+            var start = DateTime.UtcNow;
+            var markerFile = System.Environment.ExpandEnvironmentVariables(MarkerFilePath);
+            while (start.Add(timeout) >= DateTime.UtcNow)
```
Comment (davidebbo) [inline Kudu.Core/Helpers/PostDeploymentHelper.cs]: So if we go over that  time, you just stop updating the marker file, but that doesn't affect the ongoing operation, right? I guess it just means that at that point you're letting DWAS kill it? Though if feels a bit strange that in that case, you don't delete the file and just let it go stale.

Maybe I'm misreading this?

File: Kudu.Core/SiteExtensions/SiteExtensionManager.cs [hunk 2]
```diff
@@ -210,9 +211,20 @@ public SiteExtensionManager(IContinuousJobsManager continuousJobManager, ITrigge
...
         public async Task<SiteExtensionInfo> InstallExtension(string id, string version, string feedUrl, SiteExtensionInfo.SiteExtensionType type, ITracer tracer, string installationArgs = null)
         {
+            var installationTask = InstallExtensionCore(id, version, feedUrl, type, tracer, installationArgs);
+
+#pragma warning disable 4014
+            // Track pending task
+            PostDeploymentHelper.TrackPendingOperation(installationTask, TimeSpan.Zero);
+#pragma warning restore 4014
+
+            return await installationTask;
```
[the remark this answers is not in the document]
Comment (suwatch) [reply Kudu.Core/SiteExtensions/SiteExtensionManager.cs]: Ok.

File: Kudu.Core/Helpers/PostDeploymentHelper.cs (hunk 1 above)
Comment (suwatch) [reply Kudu.Core/Helpers/PostDeploymentHelper.cs]: Correct.   This PR has **no** effect on current behavior.   It compliments by adding pending marker file.

File: Kudu.Core/Helpers/PostDeploymentHelper.cs [hunk 3]
```diff
@@ -394,6 +394,51 @@ public static void RunPostDeploymentScripts(TraceListener tracer)
...
+        /// <summary>
+        /// As long as the task was not completed, we will keep updating the marker file.
+        /// The routine completes when either task completes or timeout.
+        /// If task is completed, we will remove the marker.
+        /// If timeout, we will leave the stale marker.
+        /// </summary>
+        public static async Task TrackPendingOperation(Task task, TimeSpan timeout)
+        {
+            const int DefaultTimeoutMinutes = 30;
+            const int DefaultUpdateMarkerIntervalMS = 10000;
```
[the remark this answers is not in the document]
Comment (suwatch) [reply Kudu.Core/Helpers/PostDeploymentHelper.cs]: The caller can pass explicit timeout and DefaultTimeoutMinutes will not be used.   For update pending marker interval (10s), it is not overridable.

[the remark this answers is not in the document]
Comment (davidebbo) [reply Kudu.Core/Helpers/PostDeploymentHelper.cs]: Yes, I was refrring to `DefaultUpdateMarkerIntervalMS`

[the remark this answers is not in the document]
Comment (suwatch) [reply Kudu.Core/Helpers/PostDeploymentHelper.cs]: Correct, it is NOT overridable.  So you suggest Default prefix may not be proper, right?   I can take it out (it becomes UpdateMarkerIntervalMS).

File: Kudu.Core/Helpers/PostDeploymentHelper.cs (hunk 1 above)
Comment (davidebbo) [reply Kudu.Core/Helpers/PostDeploymentHelper.cs]: As discussed, it's best to still clean up the file in that case

Comment (suwatch) [reply Kudu.Core/Helpers/PostDeploymentHelper.cs]: will do

**Comment by suwatch** [comment]
commit 790af48

**Pull request closed without merging**